dwave/embedding/transforms.py

```diff
@@ -14,48 +14,293 @@
+            {(0, 1): 1.0, (0, 3): 1.0, (1, 2): 1.0, (2, 3): -1.0}
+
+
+        See also:
+            :func:`.embed_ising`, :func:`.embed_qubo`
+
+        """
+        return_vartype = source_bqm.vartype
+        if smear_vartype is dimod.SPIN:
+            source_bqm = source_bqm.spin
+        elif smear_vartype is dimod.BINARY:
+            source_bqm = source_bqm.binary
+
+        # create a new empty binary quadratic model with the same class as
+        # source_bqm if it's shapeable, otherwise use AdjVectorBQM
+        if source_bqm.shapeable():
+            target_bqm = source_bqm.base.empty(source_bqm.vartype)
+        else:
+            target_bqm = dimod.AdjVectorBQM.empty(source_bqm.vartype)
+
+        # add the offset
+        target_bqm.add_offset(source_bqm.offset)
+
+        if isinstance(chain_strength, abc.Mapping):
+            strength_iter = (chain_strength[v] for v in source_bqm.linear)
+        else:
+            strength_iter = itertools.repeat(chain_strength)
+
+        offset = 0
+        # spread the linear source bias equally over the target variables in the
+        # chain and add chain edges as necessary
+        for (v, bias), strength in zip(source_bqm.linear.items(), strength_iter):
+            chain = self.get(v)
+            if chain is None:
+                raise MissingChainError(v)
+
+            #we check that this is nonzero in __init__
+            b = bias / len(chain)
+
+            target_bqm.add_variables_from({u: b for u in chain})
+
+            if len(chain) == 1:
+                # in the case where the chain has length 1, there are no chain
+                # quadratic biases, but we none-the-less want the chain
+                # variables to appear in the target_bqm
+                q, = chain
+                target_bqm.add_variable(q, 0.0)
+            elif target_bqm.vartype is dimod.SPIN:
+                for p, q in self.chain_edges(v):
+                    target_bqm.add_interaction(p, q, -strength)
+                    offset += strength
+            else:
+                # this is in spin, but we need to respect the vartype
+                for p, q in self.chain_edges(v):
+                    target_bqm.add_interaction(p, q, -4*strength)
+                    target_bqm.add_variable(p, 2*strength)
+                    target_bqm.add_variable(q, 2*strength)
+
+        target_bqm.add_offset(offset)
+
+        # next up the quadratic biases, spread the quadratic biases evenly over
+        # the available interactions
+        for (u, v), bias in source_bqm.quadratic.items():
+            interactions = list(self.interaction_edges(u, v))
+
+            if not interactions:
+                raise MissingEdgeError(u, v)
+
+            b = bias / len(interactions)
+
+            target_bqm.add_interactions_from((u, v, b) for u, v in interactions)
+
+        if return_vartype is dimod.BINARY:
+            return target_bqm.binary
```

Comment:
In the previous version of embed_bqm(), I was getting errors when running the dimod parametrized tests (dimod/testing/sampler.py) (called in test_embedding_composite.py), because we were passing a VartypeView around and trying to change its vartype at some point. 

I think returning target_bqm.binary as opposed to target_bqm.change_vartype(‘BINARY’, inplace=False) might cause the same problem. (Same for spin below)

Reply:
oh dang, thanks for the heads up!

Reply:
Doesn't help now, but for the future... https://github.com/dwavesystems/dimod/issues/685

Reply:
Oh!  Looking at that code more closely, it's doing something quite silly -- `target_bqm` is already being constructed with the right vartype, so that final conversion is superfluous.